dep-types-inference/dep-types.md

```diff
@@ -0,0 +1,687 @@
+## Preface
+
+At first, I was modelling a functional language with `TypeApplication`, `RankNTypes` and `Impredicativity`. But at some point I scrapped and recycled the idea in favour of a language with dependent types. The reason was simple - to prevent duplication of code and reduce the count of namespaces to handle. This, however, had consequences, but more on that later.
+
+
+## Language
+
+Okay, lets describe the language we model.
+
+I will use the following convention:
+
+- `A`, `B`, `C`  - programs;
+- `f`, `g` - field names;
+- `T`, `U` - types;
+- `K` - kinds;
+- `x`, `y` - variables.
+
+The programs, types and kinds in our language are the same entity, but for the sake of clarity in certain places it is useful to think of them as "types" and "kinds".
+
+We have:
+
+<table>
+  <tr>
+    <th>
+      Syntax
+    </th>
+    <th>
+      Explanation
+    </th>
+  </tr>
+  <tr>
+    <td>
+      <pre>
+        x, y, z
+      </pre>
+    </td>
+    <td>
+      variables
+    </td>
+  </tr>
+  <tr>
+    <td>
+      <pre>
+        A B
+      </pre>
+    </td>
+    <td>
+      application (function calls)
+    </td>
+  </tr>
+  <tr>
+    <td>
+      <pre>
+        fun (x : T) -> B
+      </pre>
+    </td>
+    <td>
+      lambdas (function literals)
+    </td>
+  </tr>
+  <tr>
+    <td>
+      <pre>
+        pi (x : T) -> B x
+      </pre>
+    </td>
+    <td>
+      dependent function types (result type can depend on parameter)
+    </td>
+  </tr>
+  <tr>
+    <td>
+      <pre>
+        Type
+      </pre>
+    </td>
+    <td>
+      the type of all types (including the `Type` itself)
+    </td>
```

Comment:
Do we have a sort of `TypeInType`? Is it for simplicity? I would emphasise this moment in more detail.

Reply:
Yes, exactly. Otherwise I have to use either additional `Box` or a hierarchy, which requires `Nat`-s and their arithmetics to be somehow present.